pkg/pgmodel/Readme.md

```diff
@@ -36,46 +80,51 @@ type Sample {
 
-                  new req
-                     v
-                 +--------+        +-------------+
-                 | Ingest | <----> | Label Cache |
-                 +--------+        +-------------+
-                  /     \
-                 /       \
- +----------------+     +----------------+       +-----------------+
- | metric handler | ... | metric handler | <===> | DB (series IDs) |
- +----------------+     +----------------+       +-----------------+
-                 \       /
-                  \     /
-                   \   /
-                    \ /
- +----------------+  |  +----------------+       +------------------+
- | insert handler | ... | insert handler | ====> | DB (metric data) |
- +----------------+     +----------------+       +------------------+
-```
+If asyncAcks=true, the dispatcher returns immediately, otherwise it waits for the metric batcher goroutine to return a result, and returns that in response.
+
+### Metric batcher
+
+The second stage consists of the metric batcher, one goroutine for each metric name. These goroutines are created in `dispatcher.InsertTs`, but are pooled for later re-use.
+
+On the first new insert request, the metric batcher creates a metric table in the database, if it doesn't exist yet.
+
+From then onwards, when the metric batcher receives a new insert request on its channel it: 
+1. Creates an unbuffered data-transfer channel for this batcher
+2. Initiates a read request to the copier on the copier read-request channel, sending a handle to the batcher's data-transfer channel
+3. Continues to (in a loop over `select`):
+   1. accept data for the metric (from the dispatcher) on its channel, writing them into a `model.Batch`, until its maximum batch size is reached (`metrics.FlushSize`, currently 2000)
+   2. attempt to write the data it has batched (as an `ingestor.copyRequest`) into the unbuffered data-transfer channel to the copier
+4. When the write into the unbuffered data-transfer channel succeeds, it begins a new batch, going back to step 2.
+
+### Copier
+
+The final stage consists of a fixed number of goroutines, by default equal to the number of database CPU cores. The copiers are started by the dispatcher, and each hold a handle to the read-request channel, which the batcher writes requests into.
+
+The copier:
 
-`TODO` metric table creation
+1. Reads up to `metrics.MaxInsertStmtPerTxn` (currently 100) read requests from its handle to the read-request channel, timing out after 20 milliseconds.
+2. Reads all of the `ingestor.copyRequest` from all of the read requests' data-transfer channels into its own batch (note: each `copyRequest` must be for a different metric).
```

Comment:
Worth mentioning that each `copyRequest` is basically one buffer(one batch).

Reply:
I haven't mentioned this. I'm not sure how relevant it is at this level.